quick-start with docker-compose

ddragosd opened the pull request:
Initial issue ref: https://github.com/openwhisk/openwhisk/pull/1312 

- [x] - Pull OpenWhisk project from GitHub, build it, and start it locally
- [x] - Setup an automated daily Travis build

rabbah commented:
Having a single place to change `localhost` to a different host would be convenient.

rabbah commented on docker-compose/README.md:
````diff
@@ -0,0 +1,96 @@
+To remove the stopped containers, clean the database files and the temporary files use:
+ 
+ ```bash
+ make destroy
````
perhaps a make clean to remove the download openwhisk tar ball?

ddragosd replied:
besides `make destroy` or rename it into `make clean` , WDYT ?

rabbah replied:
destroy removes `rm -rf ~/tmp/openwhisk` but not `openwhisk-master.tar.gz` and `openwhisk-master` in the local checkout. Or did I miss it?

ddragosd replied:
👍 I'm adding the logic to remove the `openwhisk-master` too.

ddragosd replied:
done.

rabbah commented:
I've tried this before and it worked for my setup (docker machine on mac) but having trouble with the latest version - will try from a linux vm.

Scanned the changes and looks 👍

rabbah commented:
General git comment - would be nice to squash and remove the merge commits so we can preserve a linear git history. If you'd like me to commit, I can do that.

ddragosd commented:
> having trouble with the latest version

@rabbah are there issues with the first hello-world function invocation or with starting docker-compose ?

rabbah commented:
```
until $(curl --output /dev/null --silent --head --fail http://192.168.99.100:5984/_all_dbs); do printf '.'; sleep 5; done
.ERROR: Service "version" doesn't have any configuration options. All top level keys in your docker-compose.yml must map to a dictionary of configuration options.
......................................^Cmake: *** [init-couchdb] Interrupt: 2
```

updating my docker compose version (from 1.5.0) as you advised.

ddragosd commented:
Yes, this PR needs `docker-compose` 1.6+. I'm updating the README.

rabbah commented:
updated to 1.9. Got further and failed here:

```
mkdir -p db/config/keys
cp ./openwhisk-master//ansible/files/auth.* db/config/keys
OPEN_WHISK_DB_PROVIDER=CouchDB \
	    OPEN_WHISK_DB_PROTOCOL=http \
	    OPEN_WHISK_DB_HOST=192.169.99.100  OPEN_WHISK_DB_PORT=5984 \
	    OPEN_WHISK_DB_USERNAME=whisk_admin OPEN_WHISK_DB_PASSWORD=some_passw0rd \
	    DB_SUBJECTS_DBS=subjects DB_WHISK_AUTHS=subjects \
	    PROJECT_HOME=./openwhisk-master/ \
	    db/createSubjects.sh
About to drop and recreate database 'subjects' on:
  http://192.169.99.100:5984
subjects
curl -s -k --user whisk_admin:some_passw0rd -X DELETE http://192.169.99.100:5984/subjects
curl -s -k --user whisk_admin:some_passw0rd -X PUT http://192.169.99.100:5984/subjects
Create key for guest
Create key for whisk.system
make: *** [init-couchdb] Error 7
```

ddragosd commented:
> Having a single place to change `localhost` to a different host

`localhost` is referenced in some env variables for `docker-compose` and also in the `Makefile`. 

When this differs we could have the script edit `/etc/hosts` with an entry `openwhisk.localhost` defaulting to `127.0.0.1` but if another IP is provided then use that one instead.  

Would this work ?

rabbah commented:
>When this differs we could have the script edit /etc/hosts with an entry openwhisk.localhost defaulting to 127.0.0.1 but if another IP is provided then use that one instead. Would this work ?

this will require `sudo`, no?
how about factoring out `localhost` and using an environment variable with `localhost` as a default? I set `DOCKER_HOST` for example since that's used by openwhisk already.

ddragosd commented:
@rabbah was there another CouchDB instance already running on `http://192.169.99.100:5984` ? If it was the docker container with CoudchDB couldn't start and the CouchDB already working has different credentials. I'd be interested in the output of `docker ps -a | grep couchdb` to see if the image has started correctly.

rabbah commented:
Fixed - used the wrong vm ip. Will post my patch when all done.
`./openwhisk-master//bin/wsk -v property set --namespace guest --auth `cat ./openwhisk-master//ansible/files/auth.guest` --apihost 172.17.0.1:443 -i`

note that setting namespace will fail and the command above should remove `--namespace guest`.

rabbah commented on docker-compose/Makefile:
```diff
@@ -0,0 +1,152 @@
+quick-start-pause:
+	echo "waiting for the Whisk invoker to come up ... "
+	until $$(curl --output /dev/null --silent --head --fail http://localhost:8085/ping); do printf '.'; sleep 5; done
+	sleep 30
```
is this `sleep 30` necessary?

ddragosd replied:
good q. it's a simple workaround to wait for kafka to be ready but it could be done differently. `30` doesn't guarantee that it's enough or not.

rabbah commented on docker-compose/Makefile:
```diff
@@ -0,0 +1,152 @@
+.PHONY: quick-start-info
+quick-start-info:
+	echo "$$(tput setaf 2)To invoke the function again use: $$(tput setaf 4)make hello-world$$(tput sgr0)"
+	echo "$$(tput setaf 2)To stop openwhisk use: $$(tput setaf 4)make destroy$$(tput sgr0)"
```
this comes through as:

```
> make quick-start-info
echo "$(tput setaf 2)To invoke the function again use: $(tput setaf 4)make hello-world$(tput sgr0)"
To invoke the function again use: make hello-world
echo "$(tput setaf 2)To stop openwhisk use: $(tput setaf 4)make destroy$(tput sgr0)"
To stop openwhisk use: make destroy
```

rabbah commented:
the gateway didn't come up for me.
docker logs on the failed container shows:
```
nginx: [emerg] PEM_read_bio_X509_AUX("/etc/ssl/openwhisk-cert.pem") failed (SSL: error:0906D066:PEM routines:PEM_read_bio:bad end line)
nginx: configuration file /etc/api-gateway/api-gateway.conf test failed
   ... using log level: 'warn'. Override it with -e 'LOG_LEVEL=<level>' 
nginx: [emerg] PEM_read_bio_X509_AUX("/etc/ssl/openwhisk-cert.pem") failed (SSL: error:0906D066:PEM routines:PEM_read_bio:bad end line)
```

rabbah commented on docker-compose/db/createSubjects.sh:
```diff
@@ -0,0 +1,65 @@
+URL_BASE="$OPEN_WHISK_DB_PROTOCOL://$OPEN_WHISK_DB_HOST:$OPEN_WHISK_DB_PORT"
+
+if [ "$OPEN_WHISK_DB_PROVIDER" == "CouchDB" ]; then
+    CURL_ADMIN="curl -s -k --user $OPEN_WHISK_DB_USERNAME:$OPEN_WHISK_DB_PASSWORD"
```
when this failed for me, i had to edit the curl to add `-v` to see why it failed (couldn't reach host).

rabbah commented:
Here's my patch - all the places where I replaced localhost:

```
> git diff
diff --git a/docker-compose/Makefile b/docker-compose/Makefile
index 08cf7fc..e166ba2 100644
--- a/docker-compose/Makefile
+++ b/docker-compose/Makefile
@@ -2,7 +2,7 @@ PROJECT_HOME ?= ./openwhisk-master/
 WSK_CLI ?= $(PROJECT_HOME)/bin/wsk
 
 OPEN_WHISK_DB_PROTOCOL ?= http
-OPEN_WHISK_DB_HOST ?= localhost
+OPEN_WHISK_DB_HOST ?= 172.17.0.1
 OPEN_WHISK_DB_PORT ?= 5984
 OPEN_WHISK_DB_PROVIDER ?= CouchDB
 OPEN_WHISK_DB_USERNAME ?= whisk_admin
@@ -28,7 +28,7 @@ download:
 .PHONY: quick-start-pause
 quick-start-pause:
        echo "waiting for the Whisk invoker to come up ... "
-       until $$(curl --output /dev/null --silent --head --fail http://localhost:8085/ping); do printf '.'; sleep 5; done
+       until $$(curl --output /dev/null --silent --head --fail http://172.17.0.1:8085/ping); do printf '.'; sleep 5; done
        sleep 30
 
 .PHONY: quick-start-info
@@ -47,7 +47,7 @@ run: setup start-docker-compose init-couchdb init-couchdb-actions init-whisk-cli
 .PHONY: setup
 setup:
        mkdir -p ~/tmp/openwhisk/apigateway/ssl
-       cd $(PROJECT_HOME)/ansible/roles/nginx/files/ && ./genssl.sh localhost
+       cd $(PROJECT_HOME)/ansible/roles/nginx/files/ && ./genssl.sh 172.17.0.1
        cp $(PROJECT_HOME)/ansible/roles/nginx/files/*.pem ~/tmp/openwhisk/apigateway/ssl
        cp -r $(PROJECT_HOME)/../apigateway/* ~/tmp/openwhisk/apigateway/
 
@@ -58,7 +58,7 @@ start-docker-compose:
 .PHONY: init-couchdb
 init-couchdb:
        echo "waiting for the database to come up ... "
-       until $$(curl --output /dev/null --silent --head --fail http://localhost:5984/_all_dbs); do printf '.'; sleep 5; done
+       until $$(curl --output /dev/null --silent --head --fail http://172.17.0.1:5984/_all_dbs); do printf '.'; sleep 5; done
 
        echo "initializing the database ... "
        # the folder config/keys is referenced from createSubjects.sh
@@ -94,9 +94,9 @@ init-couchdb-actions:
 .PHONY: init-whisk-cli
 init-whisk-cli:
        echo "waiting for the Whisk controller to come up ... "
-       until $$(curl --output /dev/null --silent --head --fail http://localhost:8888/ping); do printf '.'; sleep 5; done
+       until $$(curl --output /dev/null --silent --head --fail http://172.17.0.1:8888/ping); do printf '.'; sleep 5; done
        echo "initializing CLI ... "
-       $(WSK_CLI) -v property set --namespace guest --auth `cat $(PROJECT_HOME)/ansible/files/auth.guest` --apihost localhost:443 -i
+       $(WSK_CLI) -v property set --auth `cat $(PROJECT_HOME)/ansible/files/auth.guest` --apihost 172.17.0.1:443 -i
 
 .PHONY: stop
 stop:
diff --git a/docker-compose/docker-compose.yml b/docker-compose/docker-compose.yml
index 0c6104b..cc4bf84 100644
--- a/docker-compose/docker-compose.yml
+++ b/docker-compose/docker-compose.yml
@@ -78,7 +78,7 @@ services:
       DB_USERNAME: whisk_admin
       DB_PASSWORD: some_passw0rd
 
-      LOADBALANCER_HOST: localhost
+      LOADBALANCER_HOST: 172.17.0.1
       LOADBALANCER_HOST_PORT: 443
     volumes:
       - ~/tmp/openwhisk/controller/logs:/logs
@@ -117,7 +117,7 @@ services:
       DB_USERNAME: whisk_admin
       DB_PASSWORD: some_passw0rd
 
-      EDGE_HOST: localhost
+      EDGE_HOST: 172.17.0.1
       EDGE_HOST_APIPORT: 443
 
       DOCKER_REGISTRY: ""
diff --git a/docker-compose/docker-whisk-controller.env b/docker-compose/docker-whisk-controller.env
index e723ce4..c126a1b 100644
--- a/docker-compose/docker-whisk-controller.env
+++ b/docker-compose/docker-whisk-controller.env
@@ -26,5 +26,5 @@ LIMITS_ACTIONS_INVOKES_CONCURRENT=5000
 LIMITS_TRIGGERS_FIRES_PERMINUTE=60000
 LIMITS_ACTIONS_INVOKES_CONCURRENTINSYSTEM=5000
 
-WHISK_API_HOST=https://localhost
+WHISK_API_HOST=https://172.17.0.1
```

ddragosd commented:
thanks @rabbah . I'm trying to integrate with the `DOCKER_HOST` env variable. IIRC from older docker version it looks like `tcp://192.168.59.103:2375` . So I'm thinking that if this ENV VAR is set in the system to extract the IP and use it instead of `localhost`.

ddragosd commented:
> Having a single place to change localhost to a different host would be convenient.

ddragosd commented:
> Having a single place to change localhost to a different host would be convenient.

 - [x] Done. The `Makefile` looks for the `DOCKER_HOST` env var and when it's defined it uses the IP. Otherwise it uses `localhost`.

 - [x] Added Travis build badge
 - [x] added `.SILENT` as default option for `make` to avoid printing the commands

ddragosd commented:
@rabbah I've integrated your suggestions and I think this time the build should work fine on your machine, without requiring any modifications.

rabbah commented:
I rebased the pull request and tried again - worked without any changes. Thanks for the fixes!
